src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -271,6 +298,47 @@ object JoinIndexRule
     }
   }
 
+  /**
+   * The method extracts all the projection fields.
+   *
+   * @param plan The plan from which to extract projections
+   * @return A sequence of [[NamedExpression]]
+   */
+  private def collectProjections(plan: LogicalPlan): Seq[NamedExpression] = {
+    plan.collect {
+      case p: Project => p.projectList
+    }.flatten
+  }
+
+  /**
+   * The method tries to map any top level condition field to the fields present in relation.
+   * It does this by going through projections.
+   *
+   * @param projections The available projection expressions
+   * @return Some of the found expression when the condition field is found otherwise None
+   */
+  private def conditionFieldsToRelationFields(
```

Comment:
Seems the function called multiple times for cr1 & cr2. Could you try to refactor this a bit?
I think we could keep Map[Expression, Expression] instead of Seq[NamedExpression] ?